File: sample_objc/ExelbidSample/Domain/EBAdTagSupport.m
//
//  EBAdTagSupport.m
//
//  Created by s Jeong on 2022/08/16.
//

#import "EBAdTagSupport.h"
#import <AdSupport/AdSupport.h>
#import <CoreLocation/CoreLocation.h>
#import <CoreTelephony/CTTelephonyNetworkInfo.h>
#import <CoreTelephony/CTCarrier.h>
#import <AppTrackingTransparency/AppTrackingTransparency.h>
#import <sys/utsname.h>

NSString * const kIDFA = @"idfa";
NSString * const kCOPPA = @"coppa";

NSString * const kYob = @"yob";
NSString * const kGender = @"gender";
NSString * const kSegments = @"segments";

NSString * const kAppVersion = @"app_version";

NSString * const kISOCountryCode = @"iso_country_code";
NSString * const kMobileCountryCode = @"mobile_country_code";
NSString * const kMobileNetworkCode = @"mobile_network_code";
NSString * const kCarrierName = @"carrier_name";

NSString * const kCountryCode = @"country_code";
NSString * const kOsVersion = @"os_version";
NSString * const kDeviceModel = @"device_model";
NSString * const kDeviceMake = @"device_make";
NSString * const kGeoLat = @"geo_lat";
NSString * const kGeoLon = @"geo_lon";


@interface EBAdTagSupport () <CLLocationManagerDelegate>

@property (nonatomic) CLLocationManager *locationManager;
@property (nonatomic, strong) CLLocation *lastLocation;

@end

@implementation EBAdTagSupport

static EBAdTagSupport *shared = nil;

+ (instancetype)sharedProvider
{
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        shared = [[self alloc] init];
    });
    
    return shared;
}

- (instancetype)init
{
    self = [super init];

    if (self) {
        
        // Device Model
        struct utsname systemInfo;
        uname(&systemInfo);
        NSString *deviceModel = [NSString stringWithCString:systemInfo.machine encoding:NSUTF8StringEncoding];
        
        // Location
        _locationManager = [[CLLocationManager alloc] init];
        
        [self locationStart];
        
        CLLocation *existingLocation = _locationManager.location;
        if (existingLocation && existingLocation.horizontalAccuracy > 0) {
            [self updateLastLocation:existingLocation];
        }

        // Avoid processing location updates when the application enters the background.
        [[NSNotificationCenter defaultCenter] addObserverForName:UIApplicationDidEnterBackgroundNotification object:[UIApplication sharedApplication] queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *note) {
            [self.locationManager stopUpdatingLocation];
        }];

        // Re-activate location updates when the application comes back to the foreground.
        [[NSNotificationCenter defaultCenter] addObserverForName:UIApplicationWillEnterForegroundNotification object:[UIApplication sharedApplication] queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *note) {
            [self.locationManager startUpdatingLocation];
        }];

        // Location update
        [_locationManager startUpdatingLocation];
    }

    return self;
}

- (NSString *)params
{
    NSMutableDictionary * adTagModel = [[NSMutableDictionary alloc] init];
    
    // IDFA
    [adTagModel setValue:[[ASIdentifierManager sharedManager].advertisingIdentifier UUIDString] forKey:kIDFA];
    
    // COPPA
    [adTagModel setValue:[NSNumber numberWithBool:self.coppa] forKey:kCOPPA];
    [adTagModel setValue:self.yob forKey:kYob];
    [adTagModel setValue:self.gender forKey:kGender];
    [adTagModel setValue:self.segment forKey:kSegments];
    
    // App Version
    [adTagModel setValue:NSBundle.mainBundle.infoDictionary[@"CFBundleShortVersionString"] forKey:kAppVersion];

    // Carrier
    CTCarrier * carrier = [self getCarrierInfo];
    [adTagModel setValue:carrier.isoCountryCode forKey:kISOCountryCode];
    [adTagModel setValue:carrier.mobileCountryCode forKey:kMobileCountryCode];
    [adTagModel setValue:carrier.mobileNetworkCode forKey:kMobileNetworkCode];
    [adTagModel setValue:carrier.carrierName forKey:kCarrierName];
    
    // Device CountryCode
    [adTagModel setValue:[[NSLocale autoupdatingCurrentLocale] countryCode] forKey:kCountryCode];
    
    // Device Model
    [adTagModel setValue:UIDevice.currentDevice.systemVersion forKey:kOsVersion];
    [adTagModel setValue:[self getDeviceModel] forKey:kDeviceModel];
    [adTagModel setValue:@"APPLE" forKey:kDeviceMake];
    
    [adTagModel setValue:[NSNumber numberWithDouble:self.lastLocation.coordinate.latitude] forKey:kGeoLat];
    [adTagModel setValue:[NSNumber numberWithDouble:self.lastLocation.coordinate.longitude] forKey:kGeoLon];
    
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:adTagModel options:NSJSONWritingPrettyPrinted error:&error];
    if (jsonData) {
        return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
    
    return @"";
}

- (NSString *)getDeviceModel
{
    struct utsname systemInfo;
    uname(&systemInfo);

    return [NSString stringWithCString:systemInfo.machine
                              encoding:NSUTF8StringEncoding];
}

- (CTCarrier *)getCarrierInfo
{
    CTTelephonyNetworkInfo *networkInfo = [[CTTelephonyNetworkInfo alloc] init];
    NSDictionary<NSString *, CTCarrier *> *serviceProviders = [networkInfo serviceSubscriberCellularProviders];
    return [serviceProviders objectForKey:[[serviceProviders allKeys] firstObject]];
}

- (void)locationStart
{
    // CLLocationDistance
    _locationManager.delegate = self;
    _locationManager.desiredAccuracy = kCLLocationAccuracyBest;
    _locationManager.distanceFilter = 100.0;
}

- (void)updateLastLocation:(CLLocation *)location
{
    NSLog(@"GEO LAT : %@, LON : %@", [NSNumber numberWithDouble:location.coordinate.latitude], [NSNumber numberWithDouble:location.coordinate.longitude]);
    self.lastLocation = location;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidEnterBackgroundNotification object:[UIApplication sharedApplication]];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationWillEnterForegroundNotification object:[UIApplication sharedApplication]];
}

- (NSString *)toJSONString
{
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:self.adUnitInfo options:NSJSONWritingPrettyPrinted error:nil];
    return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
}

#pragma mark - <CLLocationManagerDelegate>

// iOS (14.0, *)
- (void)locationManagerDidChangeAuthorization:(CLLocationManager *)manager
{
    if (@available(iOS 14.0, *)) {
        switch (manager.authorizationStatus) {
            case kCLAuthorizationStatusNotDetermined:
                [manager requestWhenInUseAuthorization];
                break;
            case kCLAuthorizationStatusDenied:
            case kCLAuthorizationStatusRestricted:
                [manager stopUpdatingLocation];
                break;
#if __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_8_0
            case kCLAuthorizationStatusAuthorizedAlways:
            case kCLAuthorizationStatusAuthorizedWhenInUse:
#else
            case kCLAuthorizationStatusAuthorized:
#endif
                [manager startUpdatingLocation];
                break;
            default:
                break;
        }
    }
}

// iOS (4.2, 14.0)
- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
    switch (status) {
        case kCLAuthorizationStatusNotDetermined:
            [manager requestWhenInUseAuthorization];
            break;
        case kCLAuthorizationStatusDenied:
        case kCLAuthorizationStatusRestricted:
            [manager stopUpdatingLocation];
            break;
#if __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_8_0
        case kCLAuthorizationStatusAuthorizedAlways:
        case kCLAuthorizationStatusAuthorizedWhenInUse:
#else
        case kCLAuthorizationStatusAuthorized:
#endif
            [manager startUpdatingLocation];
            break;
        default:
            break;
    }
}

#pragma mark - <CLLocationManagerDelegate>

- (void)locationManager:(CLLocationManager *)manager didUpdateLocations:(NSArray *)locations
{
    NSLog(@"didUpdateLocations");
    for (CLLocation *location in locations) {
        if ([location.timestamp timeIntervalSinceDate:self.lastLocation.timestamp] < 0) {
            [self updateLastLocation:location];
        }
    }
}

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
{
    NSLog(@"didFailWithError : %@", error);
    if (error.code == kCLErrorDenied) {
        [self.locationManager stopUpdatingLocation];
    }
}

#pragma mark - <CLLocationManagerDelegate> (iOS < 6.0)

- (void)locationManager:(CLLocationManager *)manager didUpdateToLocation:(CLLocation *)newLocation fromLocation:(CLLocation *)oldLocation
{
    NSLog(@"didUpdateToLocation");
    if ([newLocation.timestamp timeIntervalSinceDate:self.lastLocation.timestamp] < 0) {
        [self updateLastLocation:newLocation];
    }
}

@end

File: sample/ExelbidSample/Domain/EBAdTagSupport.swift
//
//  AdSupportManager.swift
//  com.zionbi.ExelbidSample-Swift
//
//  Created by Jaeuk Jeong on 2023/07/26.
//

import Foundation
import UIKit
import WebKit
import AdSupport
import CoreLocation
import CoreTelephony

struct EBAdTagModel : Encodable {

    var idfa: String?

    var coppa: Bool?
    var yob: String?
    var gender: String?
    var segment: [String : String]?
    
    // 앱 버전
    var app_version: String?
    
    // Carrier
    var iso_country_code: String?
    var mobile_country_code: String?
    var mobile_network_code: String?
    var carrier_name: String?
    
    // Device
    var country_code: String?
    var os_version: String?
    var device_model: String?
    var device_make: String = "APPLE"
    
    // Location
    var geo_lat: Double?
    var geo_lon: Double?
}

class EBAdTagSupport : NSObject, CLLocationManagerDelegate{
    
    private var locationManager = CLLocationManager()
    private var lastLocation: CLLocation?

    static var shared = EBAdTagSupport()
    
    var coppa: Bool?
    var yob: String?
    var gender: String?
    var segment: [String : String]?
    
    var params: String {
        var adTagModel = EBAdTagModel()
        
        adTagModel.idfa = ASIdentifierManager.shared().advertisingIdentifier.uuidString
        adTagModel.coppa = self.coppa ?? true
        adTagModel.yob = self.yob
        adTagModel.gender = self.gender
        adTagModel.segment = self.segment
        
        adTagModel.app_version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
        
        adTagModel.os_version = UIDevice.current.systemVersion
        adTagModel.device_model = getDeviceModel()
        adTagModel.device_make = "APPLE"
        
        if let lastLocation = self.lastLocation {
            adTagModel.geo_lat = lastLocation.coordinate.latitude
            adTagModel.geo_lon = lastLocation.coordinate.longitude
        }
        
        if let carrier = self.getCarrierInfo() {
            adTagModel.iso_country_code = carrier.isoCountryCode ?? ""
            adTagModel.mobile_country_code = carrier.mobileCountryCode
            adTagModel.mobile_network_code = carrier.mobileNetworkCode
            adTagModel.carrier_name = carrier.carrierName?.replacingOccurrences(of: " ", with: "")
        }
        
        guard let jsonData = try? JSONEncoder().encode(adTagModel) else {
            return "{}"
        }
        
        return String(data: jsonData, encoding: .utf8)!
    }
    
    override init() {
        super.init()
        
        // 위치 사용
        self.locationStart()
    }
    
    func getDeviceModel() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let model = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        
        return model
    }
    
    func getCarrierInfo() -> CTCarrier? {
        var carrier: CTCarrier?
        if #available(iOS 12.0, *) {
            if let carriers = CTTelephonyNetworkInfo().serviceSubscriberCellularProviders {
                for (_, value) in carriers {
                    if value.isoCountryCode != nil {
                        carrier = value
                    }
                }
            }
        } else{
            carrier = CTTelephonyNetworkInfo().subscriberCellularProvider
        }
    
        return carrier
    }
    
    func locationStart() {
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyBest
        locationManager.distanceFilter = kCLDistanceFilterNone
        locationManager.pausesLocationUpdatesAutomatically = true
        
        locationManager.startUpdatingLocation()
    }

    // MARK: - CLLocationManagerDelegate
    
    // 위치 사용 권한에 따라 값 조회
    func changeAuthorization(status: CLAuthorizationStatus) {
        switch status {
            case .authorizedWhenInUse, .authorizedAlways:
                // 앱 사용중에만, 항상
                locationManager.startUpdatingLocation()
            case .notDetermined:
                // 결정 안됨
                // 사용자에게 권한 요청
                locationManager.requestWhenInUseAuthorization()
            case .denied, .restricted:
                // 거절, 활성 제한됨
                locationManager.stopUpdatingLocation()
            default:
                break
        }
    }

    // iOS 14.0+
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        if #available(iOS 14.0, *) {
            changeAuthorization(status: manager.authorizationStatus)
        }
    }

    // Deprecated iOS 14.0
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        changeAuthorization(status: status)
    }

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        for location in locations {
            if let lastLocation = lastLocation {
                if location.horizontalAccuracy < 1 || location.timestamp.timeIntervalSince(lastLocation.timestamp) < 0 {
                    continue;
                }
            }

            self.lastLocation = location
        }
    }

    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("locationManager didFailWithError - \(error.localizedDescription)")
    }
}
